File: pkgs/ts/src/request.ts
import { ApiRequest, HttpMethod } from "./types";

export class ApiRequestBuilder {
    private request: ApiRequest;

    constructor(method: HttpMethod, path: string) {
        this.request = {
            method,
            path,
            headers: {
                "Content-Type": "application/json",
            },
        };
    }

    static get(path: string): ApiRequestBuilder {
        return new ApiRequestBuilder(HttpMethod.GET, path);
    }

    static post(path: string): ApiRequestBuilder {
        return new ApiRequestBuilder(HttpMethod.POST, path);
    }

    static put(path: string): ApiRequestBuilder {
        return new ApiRequestBuilder(HttpMethod.PUT, path);
    }

    static delete(path: string): ApiRequestBuilder {
        return new ApiRequestBuilder(HttpMethod.DELETE, path);
    }

    static patch(path: string): ApiRequestBuilder {
        return new ApiRequestBuilder(HttpMethod.PATCH, path);
    }

    withHeader(key: string, value: string): ApiRequestBuilder {
        this.request.headers = this.request.headers || {};
        this.request.headers[key] = value;
        return this;
    }

    withHeaders(headers: Record<string, string>): ApiRequestBuilder {
        this.request.headers = {
            ...this.request.headers,
            ...headers,
        };
        return this;
    }

    withBody(body: any): ApiRequestBuilder {
        this.request.body = body;
        return this;
    }

    build(): ApiRequest {
        return this.request;
    }
}

File: pkgs/rust/src/request.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum HttpMethod {
    GET,
    POST,
    PUT,
    DELETE,
    PATCH,
    HEAD,
    OPTIONS,
}

impl HttpMethod {
    pub fn as_str(&self) -> &'static str {
        match self {
            HttpMethod::GET => "GET",
            HttpMethod::POST => "POST",
            HttpMethod::PUT => "PUT",
            HttpMethod::DELETE => "DELETE",
            HttpMethod::PATCH => "PATCH",
            HttpMethod::HEAD => "HEAD",
            HttpMethod::OPTIONS => "OPTIONS",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiRequest {
    pub method: HttpMethod,
    pub path: String,
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub headers: HashMap<String, String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<serde_json::Value>,
}

impl ApiRequest {
    pub fn new(method: HttpMethod, path: impl Into<String>) -> Self {
        let mut headers = HashMap::new();
        headers.insert("Content-Type".to_string(), "application/json".to_string());

        Self {
            method,
            path: path.into(),
            headers,
            body: None,
        }
    }

    pub fn get(path: impl Into<String>) -> Self {
        Self::new(HttpMethod::GET, path)
    }

    pub fn post(path: impl Into<String>) -> Self {
        Self::new(HttpMethod::POST, path)
    }

    pub fn put(path: impl Into<String>) -> Self {
        Self::new(HttpMethod::PUT, path)
    }

    pub fn delete(path: impl Into<String>) -> Self {
        Self::new(HttpMethod::DELETE, path)
    }

    pub fn patch(path: impl Into<String>) -> Self {
        Self::new(HttpMethod::PATCH, path)
    }

    pub fn with_header(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.insert(key.into(), value.into());
        self
    }

    pub fn with_headers(mut self, headers: HashMap<String, String>) -> Self {
        self.headers.extend(headers);
        self
    }

    pub fn with_body(mut self, body: Option<serde_json::Value>) -> Self {
        self.body = body;
        self
    }
}
